naga/src/back/pipeline_constants.rs

```diff
@@ -209,60 +225,55 @@ pub fn process_overrides<'a>(
 
 fn process_pending(
     module: &mut Module,
+    original_module_types: &UniqueArena<Type>,
     override_map: &HandleVec<Override, Handle<Constant>>,
     adjusted_global_expressions: &HandleVec<Expression, Handle<Expression>>,
-) -> Result<(), PipelineConstantError> {
-    for (handle, ty) in module.types.clone().iter() {
+) {
+    for (handle, ty) in original_module_types.iter() {
         if let TypeInner::Array {
             base,
             size: crate::ArraySize::Pending(size),
             stride,
         } = ty.inner
         {
-            let expr = match size {
+            match size {
                 crate::PendingArraySize::Expression(size_expr) => {
-                    adjusted_global_expressions[size_expr]
+                    let expr = adjusted_global_expressions[size_expr];
+                    if expr != size_expr {
+                        module.types.replace(
```

Comment:
`UniqueArena::replace` is specified to panic if the new value already exists in the arena. The WGSL front end also creates `PendingArraySize::Expression` array sizes. Why do we believe that these `replace` calls will never panic?

If this isn't a bug, it definitely deserves a comment.

Reply:
The expression handle in `PendingArraySize::Expression` is the only handle to that expression.

I realize that this is not validated though, I will add some validation for this.

Reply:
> The expression handle in `PendingArraySize::Expression` is the only handle to that expression.

This is needed for type equivalency, arrays sized via an override expression should never be equal to other arrays.

Reply:
The validation is not necessary, I missed the fact that we are working with a `UniqueArena` https://github.com/gfx-rs/wgpu/pull/6787#discussion_r1929859698.